### ISSUE TITLE
Utilize a second keyboard?

### ISSUE THREAD

**Issue by tildesarecool** [Fajar-RahmadJaya/KeyTik#4]
I was wondering if this could be added (it's not mentioned as a current feature). 

I would like to use KeyTik on a secondary keyboard, leaving my primary keyboard unmodified. See LuaMacros and HIDmacros as an example of a second keyboard implementation (those are both really outdated).

**Comment by Fajar-RahmadJaya** [comment]
Hi!! Thank you for your suggestion and sorry for late response.
If what you mean is an ability to assign remap or script to multiple or different physical  keyboard, then i do some research about it. I think it's possible to do that in AutoHotkey with some library.
It's a good idea actually. But right know i am still working to refine KeyTik functionality. I also still need to improve repository and other. 
But don't worry maybe i will implement it in the future. So look forward to it.

**Comment by tildesarecool** [comment]
> Hi!! Thank you for your suggestion and sorry for the late response. If what you mean is an ability to assign remap or script to multiple or different physical keyboard, then i do some research about it. I think it's possible to do that in AutoHotkey with some library. And if it's possible in AutoHotkey then that mean it's also possible for KeyTik to do it. I think it's a good idea. But right know I am still working to refine KeyTik functionality. I also still need to improve repository and other. But don't worry maybe I will implement it in the future. So look forward to it.

Yes, I did mean physical keyboard. When I made my suggestion I didn't really understand how your script works (I may not still). It looks like Python is used to create the GUI and a text file with an AHK extension and AHK handles the macros execution. I don't know how close that is to accurate. I thought you would either write the whole thing in AutoHotkey (I think it has a GUI library?) or do the whole thing in python since it has PyAutoGUI and other libraries to do with AutoHotkey does. I'm not criticizing, I just didn't realize you wrote it the way you did.

Anyway, I was doing some research on what it would take to write a macro program (inspired by HIDMacros and/or LuaMacros) and found a few things out. Namely that Python can test for and utilize different keyboards via the VID and/or PID strings. 

I don't know if AutoHotkeys can grab VID/PID strings from plugged in HID devices or not. Nor do I know if the Python script could grab it and utilize that into the Authotkey script. The utility USBDeview by NirSoft (https://www.nirsoft.net/utils/usb_devices_view.html) helped me when I was researching if any of this sounds interesting.

Sorry for the long winded reply. All the research just happened to be fresh in my mind. You can close this as a suggestion/issue.

**Comment by tildesarecool** [comment]
Just a random idea, not important.

**Comment by Fajar-RahmadJaya** [comment]
Your suggestion is very important!! The idea to assign remap on different physical keyboard is a very good idea if you ask me. Not many remapper have this feature after all, it might be one of the unique feature for KeyTik that make it standout then the rest of remapper. I already have idea on my mind to what the UI will be or how to do it. I might add this feature after this release or in this release as experimental. As for how KeyTik work it's like what you describe but i use only python and AutoHotkey is just for macro excutor not UI. My reason for this is in my opinion AutoHotkey have rich feature but complex use, so why not make a tool to make it easier (also it much easier, simple and lightweight this way). Why it have 0,2 AutoHotkey language is just an imported profile for auto clicker, screen clicker and multiple files opener i include on the download (user can still deleted it) not an UI or anything. How it work is as simple as this :
* Using python for the UI.
* The UI is to take input from user then it will create an AutoHotkey script.
* There is 2 mode, default mode which is just simple to take input like original key and remap key, shortcut and text mode that can made AutoHotkey script as a whole for more advance scripting.
* With that input, KeyTik will make an AutoHotkey file with (.ahk) extension in it (you can see it in folder : KeyTik\_internal\Data\Active) 
  * Example, if you input 'e' in default key and 'r' in remap key and assign shortcut 'q & e' (change 'e' to 'r' with shortcut 'q+e') the AutoHotkey script will look like this :

```
; default (to tell the mode)
^!p:: ; Syntax to exit script
ExitApp ; Syntax to exit script

toggle := false ; Syntax to make shortcut work

~q & e::toggle := !toggle ; The shortcut will goes to this

#If toggle ; Syntax to make shortcut work

e::r ; The remap or script will goes to this part

#If ; Syntax to make shortcut work
```


* Then AutoHotkey can execute the code that has been made from the UI (you can also run it with just double click on the AutoHotkey script)

To make it simple, python is just to make GUI, functionality, and made AutoHotkey Script and the one that will make it work is AutoHotkey. Basically its a program to make AutoHotkey code much easier for basic use while can still use it for advance use. This give me an idea to include 'How KeyTik Work' on Readme, maybe i will add it sometimes.

As for feature to assign remap on another physical keyboard, it also like what you describe using keyboard VID and/or PID. after my previous research what i have in mind is like this :
* I still don't know if AutoHotkey have built in feature to do this (I will research it further when i work on this feature) but there is library or some sort for AutoHotkey to do this (for now what i find is in: https://github.com/evilC/AutoHotInterception [I will research it further if i find better way i will use that)
* From that tool, what i capture is it can assign script for multiple keyboard using keyboard ID.
* And yes, look like there is python library that can do that (for now what i find is: https://github.com/rene-aguirre/pywinusb [Again i will research it further if there is better way])
* So my implementation idea for KeyTik for now is :
  * A button or entry below script name to take keyboard ID automatically so it can be used in the text mode or default mode so it can also be used for imported script.
  * it will give user a choice what keyboard they ant to assign with the script.
  * Then KeyTik will take the keyboard ID showed in the entry.
  * After all the input finished and finish button clicked then it will write a necessary script to work with keyboard id as input.
  * The script created and can be used by all functionality KeyTik have.

For now that is my idea. I almost done with current release and will work on that after this. But again i don't know if it will work if i yet try it but the idea is there and maybe have high chance for success. 

I really appreciate your suggestion, it give me a lot of idea for improvement. So once again, thank you for your suggestion. Also i ask your permission to reopen the issue until it implemented or reach a dead end. Sorry for the long reply, i hope it's easy to understand.

**Comment by Fajar-RahmadJaya** [comment]
Thanks a lot! I’ll keep you updated in this issue once it’s done or if anything changes.

**Comment by Fajar-RahmadJaya** [comment]
Hi @tildesarecool !!
Finally it done! After some experimenting, i decide to use AutoHotkey wrapper named AutoHotkey Interception by evilC. AutoHotkey Interception is using Interception Driver by oblitum to make it work. So KeyTik also need to install Interception Driver first to be able to work. I already try it using only my laptop internal keyboard and external mouse and it work. But i still don't know whether it will work on different device because i don't have another keyboard to try it. 

I still have some question about testing and maybe i would not close this issue until i am sure it work. I will try another way to really test it. But if you don't mind can you also help me test it to answer this following question?

Does windows showing warning when we run KeyTik.exe?
Does assign profile on specific keyboard feature is working on different keyboard?

If you want to try it, go to [KeyTik v1.3 release](https://github.com/Fajar-RahmadJaya/KeyTik/releases/tag/v1.3) for more detailed description, install guide and download.

Thank you!!

**Comment by tildesarecool** [comment]
I don't know if I'm going to be a good tester on this as I've disabled the safety exe warnings on my system (which no one should ever do, _ever_). I'm assuming your exe is not signed so anyone running the default/sane settings of Windows will get a "this exe isn't signed" warning from Windows (and/or a UAC prompt). Which isn't a bad thing. 

I installed the driver monitor with AHK following those instructions and it is recognizing my secondary keyboard assignment. I haven't yet managed to assign an action to a key combination yet, that will be my next test. Not sure when that will be. 

By the way, for it's worth I think you've done a really job on the code and documentation of your script on this repository. I'm not really a developer so that may not mean anything. But it seems very complete to me.

**Comment by Fajar-RahmadJaya** [comment]
Thank you for testing it!
</br>

> I'm assuming your exe is not signed so anyone running the default/sane settings of Windows will get a "this exe isn't signed" warning from Windows (and/or a UAC prompt). Which isn't a bad thing.

Yes, this is what i am worried about. Because there might be someone who still worried if this warning pop up. I want to make sure they can assure that it's safe. I have ask one person to try it and he say it didn't show warning. Still, one test is not enough. I will ask another person to try it later. 
</br>

> I installed the driver monitor with AHK following those instructions and it KeyTik is recognizing my secondary keyboard assignment. I haven't yet managed to assign an action to a key combination yet with KeyTik, that will be my next test. Not sure when that will be.

I'm glad to hear it. If it can remap or recognizing your secondary keyboard then it should work for the rest of feature. This should be sufficient for test. But let me know if you managed to assign an action to a key combination. Also i will open this issue for a while until i really satisfied with this version. So if you encounter any bug or error, feel free to let me know. Thank you for testing it!
</br>

> By the way, for it's worth I think you've done a really job on the code and documentation of your script on this repository. I'm not really a developer so that may not mean anything. But it seems very complete to me.

Thank you!! As i maintain my repository i also learn a lot of thing in it's way. I also still have a lot of plan, especially for documentation. So i hope that will improve the documentation more.
